File: Sort Vowels in a String/Sort Vowels in a String.py
# Solution for Sort Vowels in a String in PY

class Solution:
    def sortVowels(self, s: str) -> str:
        # Helper function to check if a character is a vowel
        def isVowel(ch: str) -> bool:
            return ch.lower() in "aeiou"

        # Step 1: Count frequency of each vowel (case-sensitive)
        vowelFrequency = {}
        for ch in s:
            if isVowel(ch):
                vowelFrequency[ch] = vowelFrequency.get(ch, 0) + 1

        # Step 2: Define vowels in sorted order (uppercase before lowercase)
        sortedVowelOrder = "AEIOUaeiou"

        # Step 3: Replace vowels in the string with sorted vowels
        s_list = list(s)  # convert to list since strings are immutable
        vowelIndex = 0

        for i in range(len(s_list)):
            if isVowel(s_list[i]):
                # Find the next available vowel from the frequency map
                while vowelFrequency.get(sortedVowelOrder[vowelIndex], 0) == 0:
                    vowelIndex += 1

                # Replace and decrement frequency
                s_list[i] = sortedVowelOrder[vowelIndex]
                vowelFrequency[sortedVowelOrder[vowelIndex]] -= 1

        return "".join(s_list)  # Final string with vowels sorted

File: Sort Vowels in a String/Sort Vowels in a String.cpp
// Solution for Sort Vowels in a String in CPP

class Solution {
public:
    string sortVowels(string s) {
        // Helper lambda to check if a character is a vowel
        auto isVowel = [](char ch) {
            char lowerChar = tolower(ch);  // Convert to lowercase for uniform comparison
            return (lowerChar == 'a' || lowerChar == 'e' ||
                    lowerChar == 'i' || lowerChar == 'o' ||
                    lowerChar == 'u');
        };

        // Step 1: Count frequency of each vowel (case-sensitive)
        unordered_map<char, int> vowelFrequency;
        for (char ch : s) {
            if (isVowel(ch)) {
                vowelFrequency[ch]++;
            }
        }

        // Step 2: Define vowels in sorted order (uppercase before lowercase)
        string sortedVowelOrder = "AEIOUaeiou";

        // Step 3: Replace vowels in the string with sorted vowels
        int vowelIndex = 0; // Pointer to current position in sortedVowelOrder
        for (int i = 0; i < s.length(); i++) {
            if (isVowel(s[i])) {
                // Find the next available vowel from the frequency map
                while (vowelIndex < sortedVowelOrder.size() &&
                       vowelFrequency[sortedVowelOrder[vowelIndex]] == 0) {
                    vowelIndex++;
                }

                // Replace and decrement frequency
                s[i] = sortedVowelOrder[vowelIndex];
                vowelFrequency[sortedVowelOrder[vowelIndex]]--;
            }
        }

        return s; // Final string with vowels sorted
    }
};
